### PR TITLE
union: enable ListR when upstreams contain local

### PR DESCRIPTION
**Pull request by Max-Sum**
#### What is the purpose of this change?

Enable fast list functions for union backend when:
- at least one of the upstreams supports fast list
- upstreams only consist of backends that support fast list and local backend.

#### Was the change discussed in an issue or in the forum before?

#3000

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

File: backend/union/union.go [hunk 1]
```diff
@@ -616,7 +637,18 @@ func (f *Fs) ListR(ctx context.Context, dir string, callback fs.ListRCallback) (
...
 			return nil
-		})
+		}
+		do := u.Features().ListR
+		if do != nil {
+			err = do(ctx, dir, callback)
+		} else {
+			// if we're here, we're gonna do a standard recursive traversal
+			list := walk.NewListRHelper(callback)
+			err = f.recurse(ctx, dir, list)
```
Comment (ncw) [inline backend/union/union.go:647]: This appears to be recursing `f` - shouldn't that be `u`?

File: backend/union/union.go [hunk 2]
```diff
@@ -616,7 +637,18 @@ func (f *Fs) ListR(ctx context.Context, dir string, callback fs.ListRCallback) (
...
 			mutex.Unlock()
 			return nil
-		})
+		}
+		do := u.Features().ListR
+		if do != nil {
+			err = do(ctx, dir, callback)
+		} else {
+			// if we're here, we're gonna do a standard recursive traversal
+			list := walk.NewListRHelper(callback)
```
Comment (ncw) [inline backend/union/union.go:646]: Why not this? Saves having to have your `recurse` function?

```go
err := walk.ListR(ctx, u, dir, true, -1, walk.ListAll, callback)
```

Comment (Max-Sum) [reply backend/union/union.go]: My fault. I did not see this function and just copy implementation from cache.

**Comment by ncw** [comment]
This looks great now - thank you :-)

I'll merge it now

**Pull request merged**